### PR TITLE
Port legacy chisel code to chisel3 for amba

### PR DESCRIPTION
**Pull request by tianrui-wei**
Signed-off-by: Tianrui Wei <tianrui@tianruiwei.com>
**Related issue**: 

**Type of change**: bug report | feature request | other enhancement

**Impact**: no functional change | API addition (no impact on existing code) | API modification

**Development Phase**: proposal |  implementation

**Release Notes**

### CODE REVIEW

**Comment by tianrui-wei** [comment]
Supercedes #3057. Basically the original PR with the tilelink changes stripped. Tilelink will be re-done separately in another PR

**Comment by sequencer** [comment]
opps, plz make the following PRs to be small ;p

**Comment by tianrui-wei** [comment]
Hi @sequencer I agree having small PRs would be a better practice in general, and I try to keep PRs as small as possible (one per directory of change). However, there're several reasons why I think bigger PRs like this one are necessary in the case of porting rocket to chisel3.

1. Most of the changes are trivial and require no manual intervention. I use a regular expression to convert `BOOL(Input())` to `Input(Bool())`, and things like this. One might argue there are not significant benefit to split these changes over a dozen PRs. It pollutes the commit history and makes it more difficult to bisect problems in the face of an issue
2. Some defines are used in tilelink modules and amba modules as well. These need to be changed as well in order to pass CI.
3. I have performed some sequential equivalence checking on some typical configurations like SmallConfig or DefaultConfig or `WithJtagDTMSystem,DefaultConfig` with jaspergold, and no functional changes are reflected. The top level module for equivalance checking is `ExampleRocketSystem`. In other words, I'm more confident in that the changes are not disruptive based on formal equivalence checking than depending on ci runs or manual reasoning (although not passing in ci would be worrying)

This leads me to believe it *might* be better to do several PRs over different directories rather than doing many small PRs, but of course we could also discuss it in the next rocket chip meeting.

cc @jerryz123

**Comment by tianrui-wei** [comment]
@sequencer Per our disucssion yesterday, could you create a staging branch for all the chisel3 changes?

**Comment by sequencer** [comment]
sorry for the delay, I was out of office these todays. Will do it 8pm cst today.

**Comment by tianrui-wei** [comment]
Thanks Jiuyang, let me know if you want any changes in the branch

**Review by sequencer — CHANGES_REQUESTED** [review]

File: src/main/scala/amba/apb/Bundles.scala [hunk 1]
```diff
@@ -2,46 +2,28 @@
...
-        prdata  := UInt(0)
-        pduser :<= BundleMap()
-      case OUTPUT =>
-        pwrite  := Bool(false)
-        paddr   := UInt(0)
-        pprot   := APBParameters.PROT_DEFAULT
-        pwdata  := UInt(0)
-        pstrb   := UInt(0)
-        pauser :<= BundleMap()
-      case _ =>
```
Comment (sequencer) [inline src/main/scala/amba/apb/Bundles.scala:42]: why removing this.

File: src/main/scala/amba/axi4/Fragmenter.scala [hunk 2]
```diff
@@ -58,16 +58,16 @@ class AXI4Fragmenter()(implicit p: Parameters) extends LazyModule
 
       /* Returns the number of beats to execute and the new address */
       def fragment(a: IrrevocableIO[AXI4BundleA], supportedSizes1: Seq[Int]): (IrrevocableIO[AXI4BundleA], Bool, UInt) = {
-        val out = Wire(a)
+        val out = Wire(a.cloneType)
```
Comment (sequencer) [inline src/main/scala/amba/axi4/Fragmenter.scala:61]: chiselTypeOf

File: src/main/scala/amba/axi4/Fragmenter.scala [hunk 3]
```diff
@@ -86,23 +86,23 @@ class AXI4Fragmenter()(implicit p: Parameters) extends LazyModule
...
 
         // The number of beats-1 to execute
-        val beats1 = Mux(bad, UInt(0), maxSupported1)
+        val beats1 = Mux(bad, 0.U, maxSupported1)
         val beats = OH1ToOH(beats1) // beats1 + 1
 
         val inc_addr = addr + (beats << a.bits.size) // address after adding transfer
         val wrapMask = a.bits.bytes1() // only these bits may change, if wrapping
-        val mux_addr = Wire(init = inc_addr)
+        val mux_addr = WireInit(inc_addr)
```
Comment (sequencer) [inline src/main/scala/amba/axi4/Fragmenter.scala:105]:
```suggestion
        val mux_addr = WireDefault(inc_addr)
```

File: src/main/scala/amba/axi4/IdIndexer.scala [hunk 4]
```diff
@@ -2,13 +2,15 @@
 
 package freechips.rocketchip.amba.axi4
 
-import Chisel._
+import chisel3._
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.util._
+import chisel3.util.log2Ceil
+import chisel3.util.Cat
```
Comment (sequencer) [inline src/main/scala/amba/axi4/IdIndexer.scala:10]:
```suggestion
import chisel3.util.{Cat, log2Ceil}
```

File: src/main/scala/amba/axi4/Parameters.scala [hunk 5]
```diff
@@ -2,12 +2,13 @@
...
 package freechips.rocketchip.amba.axi4
 
-import Chisel._
 import chisel3.internal.sourceinfo.SourceInfo
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.util._
 import scala.math.max
+import chisel3.util.isPow2
+import chisel3.util.log2Up
```
Comment (sequencer) [inline src/main/scala/amba/axi4/Parameters.scala:11]:
```suggestion
import chisel3.util.{isPow2, log2Up}
```

File: src/main/scala/amba/axi4/Xbar.scala [hunk 6]
```diff
@@ -148,38 +148,42 @@ class AXI4Xbar(
...
-        when (out(i).aw.fire()) { latched := Bool(false) }
+        when (awOut(i).io.enq.fire) { latched := true.B }
+        when (out(i).aw.fire) { latched := false.B }
 
         // Block W if we do not have an AW source
         io_out(i).w.valid := out(i).w.valid && awOut(i).io.deq.valid // depends on awvalid (but not awready)
         out(i).w.ready := io_out(i).w.ready && awOut(i).io.deq.valid
         awOut(i).io.deq.ready := out(i).w.valid && out(i).w.bits.last && io_out(i).w.ready
+      } else {
+        awOut(i).io := DontCare // aw out queue is not used when outsize == 1
```
Comment (sequencer) [inline src/main/scala/amba/axi4/Xbar.scala:186]: I don't think the comment is complied with the code.

File: src/main/scala/devices/debug/Periphery.scala [hunk 7]
```diff
@@ -330,9 +324,12 @@ object Debug {
       }
 
       debug.apb.foreach { apb =>
-        apb.tieoff()
```
Comment (sequencer) [inline src/main/scala/devices/debug/Periphery.scala:333]: why remove this function call?

File: src/main/scala/tilelink/AtomicAutomata.scala [hunk 8]
```diff
@@ -2,11 +2,17 @@
...
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.util._
 import scala.math.{min,max}
+import chisel3.util.PriorityMux
+import chisel3.util.Cat
+import chisel3.util.FillInterleaved
+import chisel3.util.Mux1H
+import chisel3.util.MuxLookup
+import chisel3.util.log2Up
```
Comment (sequencer) [inline src/main/scala/tilelink/AtomicAutomata.scala:15]:
```suggestion
import chisel3.util.{PriorityMux, Cat, FillInterleaved, Mux1H, MuxLookup, log2Up}
```

File: src/main/scala/tilelink/AtomicAutomata.scala [hunk 9]
```diff
@@ -145,18 +151,18 @@ class TLAtomicAutomata(logical: Boolean = true, arithmetic: Boolean = true, conc
           Mux(a_cam_a.bits.opcode(0), logic_out, arith_out)
 
         // Potentially mutate the message from inner
-        val source_i = Wire(in.a)
+        val source_i = Wire(in.a.cloneType)
```
Comment (sequencer) [inline src/main/scala/tilelink/AtomicAutomata.scala:154]: chiselTypeof

File: src/main/scala/tilelink/Bundles.scala [hunk 10]
```diff
@@ -222,61 +225,61 @@ final class TLBundleD(params: TLBundleParameters)
...
-  def a: DecoupledIO[TLBundleA] = optA.getOrElse(Wire(Decoupled(new TLBundleA(params))))
-  def b: DecoupledIO[TLBundleB] = optB.getOrElse(Wire(Decoupled(new TLBundleB(params))))
-  def c: DecoupledIO[TLBundleC] = optC.getOrElse(Wire(Decoupled(new TLBundleC(params))))
-  def d: DecoupledIO[TLBundleD] = optD.getOrElse(Wire(Decoupled(new TLBundleD(params))))
-  def e: DecoupledIO[TLBundleE] = optE.getOrElse(Wire(Decoupled(new TLBundleE(params))))
+  def a: DecoupledIO[TLBundleA] = optA.getOrElse(WireInit(0.U.asTypeOf(Decoupled(new TLBundleA(params)))))
+  def b: DecoupledIO[TLBundleB] = optB.getOrElse(WireInit(0.U.asTypeOf(Decoupled(new TLBundleB(params)))))
+  def c: DecoupledIO[TLBundleC] = optC.getOrElse(WireInit(0.U.asTypeOf(Decoupled(new TLBundleC(params)))))
+  def d: DecoupledIO[TLBundleD] = optD.getOrElse(WireInit(0.U.asTypeOf(Decoupled(new TLBundleD(params)))))
+  def e: DecoupledIO[TLBundleE] = optE.getOrElse(WireInit(0.U.asTypeOf(Decoupled(new TLBundleE(params)))))
```
Comment (sequencer) [inline src/main/scala/tilelink/Bundles.scala:262]:
```suggestion
  def a: DecoupledIO[TLBundleA] = optA.getOrElse(WireDefault(0.U.asTypeOf(Decoupled(new TLBundleA(params)))))
  def b: DecoupledIO[TLBundleB] = optB.getOrElse(WireDefault(0.U.asTypeOf(Decoupled(new TLBundleB(params)))))
  def c: DecoupledIO[TLBundleC] = optC.getOrElse(WireDefault(0.U.asTypeOf(Decoupled(new TLBundleC(params)))))
  def d: DecoupledIO[TLBundleD] = optD.getOrElse(WireDefault(0.U.asTypeOf(Decoupled(new TLBundleD(params)))))
  def e: DecoupledIO[TLBundleE] = optE.getOrElse(WireDefault(0.U.asTypeOf(Decoupled(new TLBundleE(params)))))
```

File: src/main/scala/tilelink/ToAHB.scala [hunk 11]
```diff
@@ -2,13 +2,17 @@
...
 import freechips.rocketchip.amba._
 import freechips.rocketchip.amba.ahb._
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.util._
 import AHBParameters._
+import chisel3.util.RegEnable
+import chisel3.util.Queue
+import chisel3.util.Cat
+import chisel3.util.log2Ceil
```
Comment (sequencer) [inline src/main/scala/tilelink/ToAHB.scala:15]:
```suggestion
import chisel3.util.{RegEnable, Queue, Cat, log2Ceil}
```

File: src/main/scala/tilelink/ToAHB.scala [hunk 12]
```diff
@@ -76,19 +80,19 @@ class TLToAHB(val aFlow: Boolean = false, val supportHints: Boolean = true, val
...
+      resetState.addr  := 0.U
 
       // The stages of the combinational pipeline
       val reg  = RegInit(resetState)
-      val send = Wire(init = reg)
-      val step = Wire(init = send)
-      val next = Wire(init = step)
+      val send = WireInit(reg)
+      val step = WireInit(send)
+      val next = WireInit(step)
```
Comment (sequencer) [inline src/main/scala/tilelink/ToAHB.scala:95]:
```suggestion
      val send = WireDefault(reg)
      val step = WireDefault(send)
      val next = WireDefault(step)
```

File: src/main/scala/tilelink/ToAPB.scala [hunk 13]
```diff
@@ -2,12 +2,16 @@
...
+import chisel3._
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.amba.apb._
 import freechips.rocketchip.amba._
 import APBParameters._
+import chisel3.util.RegEnable
+import chisel3.util.Queue
+import chisel3.util.log2Ceil
+import chisel3.util.Cat
```
Comment (sequencer) [inline src/main/scala/tilelink/ToAPB.scala:14]:
```suggestion
import chisel3.util.{RegEnable, Queue, log2Ceil, Cat}
```

File: src/main/scala/tilelink/ToAXI4.scala [hunk 14]
```diff
@@ -2,16 +2,21 @@
...
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.util._
 import freechips.rocketchip.amba.axi4._
 import freechips.rocketchip.amba._
+import chisel3.util.log2Ceil
+import chisel3.util.UIntToOH
+import chisel3.util.Queue
+import chisel3.util.Decoupled
+import chisel3.util.Cat
```
Comment (sequencer) [inline src/main/scala/tilelink/ToAXI4.scala:15]:
```suggestion
import chisel3.util.{log2Ceil, UIntToOH, Queue, Decoupled, Cat}
```

File: src/main/scala/tilelink/ToAXI4.scala [hunk 15]
```diff
@@ -104,16 +109,16 @@ class TLToAXI4(val combinational: Boolean = true, val adapterName: Option[String
 
       // Construct the source=>ID mapping table
       val map = new TLtoAXI4IdMap(edgeIn.client)
-      val sourceStall = Wire(Vec(edgeIn.client.endSourceId, Bool()))
-      val sourceTable = Wire(Vec(edgeIn.client.endSourceId, out.aw.bits.id))
-      val idStall = Wire(init = Vec.fill(edgeOut.master.endId) { Bool(false) })
+      val sourceStall = WireDefault(VecInit.fill(edgeIn.client.endSourceId)(false.B))
+      val sourceTable = WireDefault(VecInit.fill(edgeIn.client.endSourceId)(0.U.asTypeOf(out.aw.bits.id)))
+      val idStall = WireInit(VecInit.fill(edgeOut.master.endId)(false.B))
```
Comment (sequencer) [inline src/main/scala/tilelink/ToAXI4.scala:114]:
```suggestion
      val idStall = WireDefault(VecInit.fill(edgeOut.master.endId)(false.B))
```

File: src/main/scala/tilelink/ToAXI4.scala [hunk 16]
```diff
@@ -145,7 +150,7 @@ class TLToAXI4(val combinational: Boolean = true, val adapterName: Option[String
       // We need these Queues because AXI4 queues are irrevocable
       val depth = if (combinational) 1 else 2
       val out_arw = Wire(Decoupled(new AXI4BundleARW(out.params)))
-      val out_w = Wire(out.w)
+      val out_w = Wire(out.w.cloneType)
```
Comment (sequencer) [inline src/main/scala/tilelink/ToAXI4.scala:153]: chiselTypeof

**Review by sequencer — CHANGES_REQUESTED** [review]
Some nitpicks

File: src/main/scala/amba/ahb/Bundles.scala [hunk 17]
```diff
@@ -2,120 +2,74 @@
...
-        hsize     := UInt(0)
-        hburst    := AHBParameters.BURST_SINGLE
-        hwrite    := Bool(false)
-        hprot     := AHBParameters.PROT_DEFAULT
-        haddr     := UInt(0)
-        hauser    :<= BundleMap()
-        hwdata    := UInt(0)
-      case _ =>
-    }
-  }
```
[the remark this answers is not in the document]
Comment (tianrui-wei) [reply src/main/scala/amba/ahb/Bundles.scala:61]: This is from a discussion between Jerry and me, he thinks `tieoff` is not good practice, is rarely used and doesn't map cleanly to chisel3 concepts. You can see later in the code that we manually tied off 2 places where it was different

File: src/main/scala/amba/ahb/Bundles.scala [hunk 18]
```diff
@@ -2,120 +2,74 @@
...
-        htrans    := AHBParameters.TRANS_IDLE
-        hsize     := UInt(0)
-        hburst    := AHBParameters.BURST_SINGLE
-        hwrite    := Bool(false)
-        hprot     := AHBParameters.PROT_DEFAULT
-        haddr     := UInt(0)
-        hauser    :<= BundleMap()
-        hwdata    := UInt(0)
-      case _ =>
-    }
```
[the remark this answers is not in the document]
Comment (tianrui-wei) [reply src/main/scala/amba/ahb/Bundles.scala:60]: ditto

File: src/main/scala/tilelink/ToAXI4.scala (hunk 16 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/ToAXI4.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/ToAXI4.scala (hunk 15 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/ToAXI4.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/ToAXI4.scala (hunk 14 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/ToAXI4.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/ToAPB.scala (hunk 13 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/ToAPB.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

Comment (tianrui-wei) [reply src/main/scala/tilelink/ToAPB.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/ToAHB.scala (hunk 12 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/ToAHB.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/ToAHB.scala (hunk 11 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/ToAHB.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/Bundles.scala (hunk 10 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/Bundles.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/AtomicAutomata.scala (hunk 9 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/AtomicAutomata.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/tilelink/AtomicAutomata.scala (hunk 8 above)
Comment (tianrui-wei) [reply src/main/scala/tilelink/AtomicAutomata.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/devices/debug/Periphery.scala (hunk 7 above)
Comment (tianrui-wei) [reply src/main/scala/devices/debug/Periphery.scala:333]: cc @jerryrzhao

File: src/main/scala/amba/axi4/Xbar.scala (hunk 6 above)
Comment (tianrui-wei) [reply src/main/scala/amba/axi4/Xbar.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/amba/axi4/Parameters.scala (hunk 5 above)
Comment (tianrui-wei) [reply src/main/scala/amba/axi4/Parameters.scala]: Thanks, addressed with https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f

File: src/main/scala/amba/axi4/IdIndexer.scala (hunk 4 above)
Comment (tianrui-wei) [reply src/main/scala/amba/axi4/IdIndexer.scala]: Thanks, addressed with [`dd3703c` (#3059)](https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f)

File: src/main/scala/amba/axi4/Fragmenter.scala (hunk 3 above)
Comment (tianrui-wei) [reply src/main/scala/amba/axi4/Fragmenter.scala]: Thanks, addressed with [`dd3703c` (#3059)](https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f)

File: src/main/scala/amba/axi4/Fragmenter.scala (hunk 2 above)
Comment (tianrui-wei) [reply src/main/scala/amba/axi4/Fragmenter.scala]: Thanks, addressed with [`dd3703c` (#3059)](https://github.com/chipsalliance/rocket-chip/pull/3059/commits/dd3703cb88a91ae4713b61be2c5e37a83fbcc77f)

File: src/main/scala/amba/apb/Bundles.scala (hunk 1 above)
Comment (tianrui-wei) [reply src/main/scala/amba/apb/Bundles.scala:42]: Same as above

File: src/main/scala/amba/axi4/Xbar.scala [hunk 19]
```diff
@@ -66,40 +66,40 @@ class AXI4Xbar(
...
       io_in(i).r.bits.id := trim(in(i).r.bits.id, r.size)
       io_in(i).b.bits.id := trim(in(i).b.bits.id, r.size)
 
       if (io_out.size > 1) {
         // Block A[RW] if we switch ports, to ensure responses stay ordered (also: beware the dining philosophers)
         val endId = edgesIn(i).master.endId
-        val arFIFOMap = Wire(init = Vec.fill(endId) { Bool(true) })
-        val awFIFOMap = Wire(init = Vec.fill(endId) { Bool(true) })
+        val arFIFOMap = WireInit(VecInit.fill(endId) { true.B })
+        val awFIFOMap = WireInit(VecInit.fill(endId) { true.B })
```
Comment (sequencer) [inline src/main/scala/amba/axi4/Xbar.scala:102]:
```suggestion
        val arFIFOMap = WireDefault(VecInit.fill(endId) { true.B })
        val awFIFOMap = WireDefault(VecInit.fill(endId) { true.B })
```

**Review by sequencer — APPROVED** [review]

**Pull request merged**